Support registration in Consul and created a driver to connect other Services Discovery

avara1986 opened the pull request:
Fixes #190 #178

Changes proposed in this PR:
- Support registration in Consul 
- Created a driver to connect other services discovery
- Refactor: added an option to services to initialize actions

avara1986 commented:
@Agalin @alexppg is it ok as a simple service discovery for Consul?

I want to add these days ACL validations and tokens :+1: 
https://www.consul.io/api-docs/acl/tokens

alexppg commented on pyms/flask/services/metrics.py:
```diff
@@ -47,14 +46,21 @@ class Service(DriverService):
         self.init_registry()
         self.serve_metrics()
 
+    def init_action(self, microservice_instance):
```
I do like this! If I do understand this correctly, with this all the logic is on the service and you don't have to touch the `create_app` file. So, you could delete https://github.com/python-microservices/pyms/blob/master/pyms/flask/app/create_app.py#L188?

Agalin replied:
I agree. Really nice change. First step for eventual 3rd party drivers too.

alexppg commented on pyms/flask/services/service_discovery.py:
```diff
@@ -0,0 +1,82 @@
+        pass
+
+
+class ServiceDiscoveryConsul(ServiceDiscoveryBase):
```
Do you plan to implement all the stages of the service discovery? It seems to be some work done, in case you don't want to implement it all:
- https://github.com/bigdatacesga/service-discovery 
- https://github.com/cablehead/python-consul

Agalin replied:
Reimplementing and maintaining the whole Consul API would be A LOT of work. While there is no official one it seems that `consulate` and `python-consul2` (fork of mentioned `python-consul` are the most up to date (consulate doesn't have any release for some time but most functions seem to work as expected).

avara1986 replied:
`consulate` is very very good implementation with a intuitive implementation, but as you said @Agalin this package hasn't a up-to-date maintenance. I need to figure out `python-consul2` :+1: 

I reimplementing because i was thinking that only need a register and deregister endpoint in our use case and other options are "over-engineering"

Agalin replied:
Personally I really like consulate. Hoping for a fork or something. 
More important part of the service discovery is accessing the registry - that can be done by consul’s dns service without a dedicated api usage. But maintaining acl tokens api as proposed - or advanced queries for consul service registry - may be a bit too complex to maintain as part of this project. So question is how advanced this service is supposed to be?

avara1986 replied:
With your approach, Could be better get access to `consulate` as a client as MVP and if we have an issue create a fork in PyMs of `consulate`, are you agree?

when i mean "client", i mean:

```
    service_discovery:
      client: "consul"
      host: "http://localhost:8500"
      autoregister: true
```

And use this service in all application with `current_app.ms.service_discovery.[all-consulate-methods]`

Agalin replied:
I think it should be under `client`.
```python
current_app.ms.service_discovery.client.[all-consulate-methods]
# generic service retrieval call provided by all implementations, would retrieve an url to access the service - to be used either by user or in other services
current_app.ms.service_discovery.find_service(name, **kwargs)
```

Agalin commented on pyms/flask/services/service_discovery.py:
```diff
@@ -0,0 +1,82 @@
+
+
+class Service(DriverService):
+    id_app = str(uuid.uuid1())
```
I'm not sure (need to check consul docs) but I think that app id should be optionally manually configured.

Agalin commented on pyms/flask/services/service_discovery.py:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    def init_action(self, microservice_instance):
+        self.client.register_service(
```
I'd argue that registration is not an obligatory step of the service discovery. While consul expects services to register themselves it may be done from consul's site. In Docker or Kubernetes env vars or DNS may be used - those don't need client-side registration action (I'm planning to implement those two to check how would it work in Docker Compose).
So the question is: should client like that implement empty registration method or instead there should be static variable that should decide if call is necessary? What do you think?

avara1986 replied:
Do you mean that could be an option in the configuration to autoregister instead to register in all cases? in example:

```yaml
    service_discovery:
      client: "consul"
      host: "http://localhost:8500"
      autoregister: true
```

Agalin replied:
Yeah. Maybe just key ‘register’ and registration params as values? No key = no registration.

Agalin commented on pyms/flask/services/service_discovery.py:
```diff
@@ -0,0 +1,82 @@
+    def get_client(self) -> ServiceDiscoveryBase:
+        client = False
+        if self.config.client == CONSUL_SERVICE_DISCOVERY:
+            client = ServiceDiscoveryConsul()
```
What about using module/class-level dict for this mapping instead? That would make it possible to define new client completely outside of the pyms project - and ease adding new mappings.

avara1986 replied:
do you mean, in example: 

```yaml
    service_discovery:
      client: "myproject.ServiceDiscovery"
      host: "http://localhost:8500"
      autoregister: true
```

I'm working this week to add a functionality to add services out of pyms, does it be a solution? or, could you put a example please?

Thanks a lot @Agalin for your feedback :)

Agalin replied:
This would be fine. Maybe with a shortcut for built-in clients?
```python
# service_discovery.py
BUILTIN_CLIENTS = {
"consul": ServiceDiscoveryConsul,
"eureka": ServiceDiscoveryEureka # example
}
```

```yaml
    service_discovery:
      # client: "consul" # uses ServiceDiscoveryConsul
      client: "myproject.ServiceDiscovery"
      host: "http://localhost:8500"
      autoregister: true
```

avara1986 replied:
Done!

Agalin commented:
ACL is nice but I wonder how would client-specific configuration options be handled? Is there any plan (maybe in tracer as there are 2 clients already) how should that be done?

avara1986 commented:
I added all the requests you did @Agalin :)

- Created a fork of `consulate`. The Version 0.6.0 isn't work correctly, and not exists new versions since 2015 (https://github.com/gmr/consulate/issues/120)
- Added a MVP of service discovery with autoregister (optional) but you can access to the client with `ms.service_discovery.client.agent.checks()` (see examples)
- Added the option to add your own Service Discovery (see examples)

Agalin commented:
Woah. There is a lot of linter fixes.

Service itself looks fine to me although I've been talking about forking (e.g. into `python-microservices` org) not vendoring `consulate`. Vendored is not useful to the community outside of pyms (and for this reason won't get much of external support), harder to provide fixes for (need whole pyms test suite to pass) and increases bundle size for people completely uninterested in service discovery (although this is a minor issue as it doesn't seem to have many dependencies).

avara1986 commented:
I moved all code into PyMS because I can't upload a new version of consulte to Pip. But... I could create a fork and create a package "pyms-consulate", upload to pip and use it as dependency, in your opinion @Agalin , is it a good solution?

Agalin commented:
That's exactly what I mean by `fork` in this context. 😄 Just so development on consulate can be continued somewhere.

alexppg commented:
+1 to forking and have a new package!

avara1986 commented:
Ready this PR! :) 

This is the fork:
https://github.com/python-microservices/consulate
And the package with last changes of consulate:
https://pypi.org/project/py-ms-consulate/

Agalin commented on pyms/flask/app/create_app.py:
```diff
@@ -58,6 +59,14 @@ def init_services(self) -> None:
                 self.services.append(service_name)
                 setattr(self, service_name, service)
+                # if getattr(service, "init_action"):
+                #     service.init_action(self)
```
Commented out code can be removed.

avara1986 replied:
Done!

Agalin commented on pyms/flask/services/service_discovery.py:
```diff
@@ -0,0 +1,72 @@
+
+CONSUL_SERVICE_DISCOVERY = "consul"
+
+EUREKA_SERVICE_DISCOVERY = "eureka"
```
It should be removed as not currently supported.

avara1986 replied:
Done!

Agalin commented on tests/test_service_discovery.py:
```diff
@@ -0,0 +1,35 @@
+        mock_request.put(url)
+        self.ms.create_app()
+
+        self.assertTrue(True)
```
What exactly is supposed to be tested here?

avara1986 replied:
If the method `self.ms.create_app()` not request the `url` with PUT Method  `mock_request raise` a assertionError

Agalin replied:
Considerinng that external library is used here, using correct API endpoint should be already tested there (in `consulate`). So if anything proper `consulate` instantiation should be tested. Anyway this assertion itself literally does nothing.

Agalin replied:
This assertion itself does nothing. Also from my experience there is no point in testing that external library (`consulate`) works as expected (i.e. calls correct endpoint on init) as it should already have it covered by tests. Instead correct call to `register_service` should be checked.

Agalin commented on tests/test_service_discovery.py:
```diff
@@ -0,0 +1,35 @@
+
+
+class ServiceDiscoveryTests(unittest.TestCase):
+    """Test common rest operations wrapper."""
```
I don't think those docstrings are correct.

avara1986 replied:
Done!

avara1986 commented:
do you see anything else @alexppg @Agalin ? is it ready to merge? :)

Agalin commented:
Yeah imo it's fine.
